fog/report/validation/src/lib.rs

```diff
@@ -1,59 +1,150 @@
+///     to create the transaction offline.
+/// (4) Take the constructed transaction to the online machine and submit to consensus.
+///
+/// Note: there is no particular reason for this to be BTreeMap instead of HashMap,
+/// except that it is slightly more portable, only requiring the alloc crate.
+pub type FogReportResponses = BTreeMap<String, ReportResponse>;
 
-/// Represents a fog public key validated to use for creating encrypted fog hints.
-/// This object should be constructed only when the IAS report has been validated.
-#[derive(Clone, Debug, Hash, Eq, PartialEq, Ord, PartialOrd)]
-pub struct IasValidatedFogPubkey(pub RistrettoPublic);
+/// A collection of unvalidated fog reports, together with an IAS verifier.
+/// This object is passed to the TransactionBuilder object.
+/// When fog is not involved, it can simply be defaulted.
+///
+/// Once constructed, this object can get validated fog pubkeys to build fog hints
+/// for transactions, without talking to the internet, and so is compatible with
+/// offline transactions to fog recipients. Only getting the FogReportResponses
+/// requires an online connection.
+#[derive(Default, Clone, Debug)]
+pub struct FogResolver {
+    responses: FogReportResponses,
+    verifier: IngestReportVerifier,
+}
 
-impl AsRef<RistrettoPublic> for IasValidatedFogPubkey {
-    fn as_ref(&self) -> &RistrettoPublic {
-        &self.0
+impl FogResolver {
+    /// Create a new FogResolver object, given serialized (unverified)
+    /// fog report server responses,
+    /// and an attestation verifier for fog ingest measurements.
+    pub fn new(responses: FogReportResponses, verifier: &Verifier) -> Self {
+        Self {
+            responses,
+            verifier: IngestReportVerifier::from(verifier),
+        }
     }
 }
 
-/// Fully resolves a public address to a fully validated fog public key structure,
-/// including all the data from the report server.
-/// This interface may include grpc and so likely cannot be implemented in a way
-/// that is safe for libmobilecoin / asynchronous java requirements.
-#[cfg_attr(any(test, feature = "automock"), automock(type Error = String;))]
-pub trait FogPubkeyResolver {
-    type Error: Display + Debug;
+impl FogPubkeyResolver for FogResolver {
     fn get_fog_pubkey(
         &self,
         recipient: &PublicAddress,
-    ) -> Result<FullyValidatedFogPubkey, Self::Error>;
+    ) -> Result<FullyValidatedFogPubkey, FogPubkeyError> {
+        if let Some(url) = recipient.fog_report_url() {
+            let url = url.to_string();
+            if let Some(result) = self.responses.get(&url) {
+                let report_id = recipient.fog_report_id().unwrap_or("").to_string();
+                for report in result.reports.iter() {
+                    if report_id == report.fog_report_id {
+                        // TODO validate x509 chain and recipient.fog_authority_sig here,
```

Comment:
this is where the x509 chain and cert checking validation code should go, after #684 is merged